### ISSUE TITLE
ERR_PNPM_LINKING_FAILED when running `pnpm install` in Docker on macOS

### ISSUE THREAD

**Issue by binaryfire** [pnpm/pnpm#7024]
### Verify latest release

- [x] I verified that the issue exists in the latest pnpm release

### pnpm version

v8.7.1

### Reproduction steps

1. Create a Node-based image which includes PNPM
2. On macOS, bind mount a project to it and attempt to run `pnpm install` a couple of times inside the container 

### Describe the Bug

This article describes the same bug I'm experiencing: https://zenn.dev/minedia/articles/2023-08-30-pnpm#%E4%BD%99%E8%AB%87%3A-pnpm%E3%81%AE%E3%82%A8%E3%83%A9%E3%83%BC

Translated:
![image](https://github.com/pnpm/pnpm/assets/5361908/5e3c1e09-d79b-4d1d-b487-50b70c158fda)

Basically, it seems like the hard links PNPM uses don't work properly under Docker for macOS.

### Expected Behavior

PNPM should work the same way in Docker containers on macOS as it does on Linux and Windows WSL2

### Which Node.js version are you using?

18.17.1

### Which operating systems have you used?

- [X] macOS
- [X] Windows
- [X] Linux

### If your OS is a Linux based, which one it is? (Include the version if relevant)

macOS 12

**Comment by binaryfire** [comment]
Another thread explaining the same issue: https://twitter.com/david_negrier/status/1608012478851743744?s=20

**Comment by weyert** [comment]
Looks like a duplicate of https://github.com/pnpm/pnpm/issues/5803 or at least related

**Comment by dotoleeoak** [comment]
I added this line in .npmrc file and it works like charm!

```properties
package-import-method=clone-or-copy
```

**Comment by AIABHISHEK** [comment]
same error occured on windows I just updated the pnpm package and it worked .
pnpm -v   8.8.0

**Comment by can-acar** [comment]
The issue still persists in Windows.
Node 18.18.0
pnpm version 8.8.0

**Comment by meister245** [comment]
The way I resolved this issue is the following:

I added a postCreateCommand and mount to my devcontainer.json

```
"postCreateCommand": "bash .devcontainer/post-create.sh",
```

In the post create script I set it to a folder in the user home:
`pnpm config set store-dir /home/vscode/.pnpm-store`

I then added a mount in the devcontainer.json file, to mount my local pnpm store in the container:

```
"mounts": [
    "type=bind,source=/Users/${localEnv:USER}/Library/pnpm/store,target=/home/vscode/.pnpm-store"
]
```

**Comment by binaryartifex** [comment]
still persisting in windows. 

```
Error: EPERM: operation not permitted, rename 'F:\[...path]\node_modules\.pnpm\cypress@13.3.2\node_modules\cypress_tmp_27736' -> 'F:\[...path]\node_modules\.pnpm\cypress@13.3.2\node_modules\cypress'
```

**Comment by hanyaonian** [comment]
each time I had this issue I have to restart my computed. pnpm version 8.9.2 & windows 10

> still persisting in windows. For some reason the cypress path is the only one that triggers this bloody error.
> 
> ```
> Error: EPERM: operation not permitted, rename 'F:\[...path]\node_modules\.pnpm\cypress@13.3.2\node_modules\cypress_tmp_27736' -> 'F:\[...path]\node_modules\.pnpm\cypress@13.3.2\node_modules\cypress'
> ```
Same here Q-Q
Every time I had this issue, I have to restart my computer (and then it works well). 
I am using pnpm version 8.9.2 and Windows 10

**Comment by darkbasic** [comment]
Same issue on Arch Linux. I'm not using containers.

**Comment by artemgordash** [comment]
> I added this line in .npmrc file and it works like charm!
> 
> ```ini
> package-import-method=clone-or-copy
> ```

with this works, thank you

**Comment by danielrotaermel** [comment]
Same issue on multiple windows machines (no docker involved). Tried pnpm `8.8.0` `8.7.0` `8.10.0` (installed via `npm i -g pnpm`). Always getting an `ERR_PNPM_LINKING_FAILED Error: EPERM: operation not permitted, rename <some random package>`

**Comment by danielrotaermel** [comment]
It seems like running Antivirus software could be the culprit https://github.com/npm/npm/issues/10826

**Comment by ickers** [comment]
FWIW, https://github.com/npm/npm/issues/10826#issuecomment-203438132 was the clue that got me past this after much thrashing @danielrotaermel (again, windows and no docker) - closed VS Code and suddenly everything worked like a charm.

**Comment by danielrotaermel** [comment]
Found a way around the file locking by the anti virus software. Rerunning pnpm
install until antivirus is done checking the files. Note: after some experimentation that I found that hoisting via copy does not work -> symlinks work. 

Maybe theres a way to increase the timeout for retries or at least log whats causing the file lock or something?

**Comment by darkbasic** [comment]
> package-import-method=clone-or-copy

That doesn't work on Linux + zfs (no containers).

**Comment by jdnavarro** [comment]
There is [copy-clone bug](https://github.com/openzfs/zfs/issues/15526) in ZFS that may be related.

In my case, setting `package-import-method=hardlink` fixes the issue.

**Comment by mtymek** [comment]
I encountered the same issue today, and cleaning `node_modules` or `pnpm prune` did not help.
Switching the "file sharing implementation" (in Docker settings) from VirtioFS to gRPC FUSE has resolved it.

<img width="853" alt="image" src="https://github.com/pnpm/pnpm/assets/777893/5fdafe15-2bef-4c96-9f77-ddefb62b353c">

**Comment by rflihxyz** [comment]
Hey, I had the same issue and managed to get this fixed in my repo

https://github.com/panoratech/Panora

The solution was to add a line "node-linker=hoisted" to the .npmrc file

read more here: https://pnpm.io/npmrc#node-linker

**Comment by gyunseo** [comment]
> ```ini
> package-import-method=clone-or-copy
> ```

@dotoleeoak thanks! that workaround works!!!

**Comment by cristian-cll** [comment]
I too tried to switch from VirtioFS to gRPC FUSE and was able to run the command without problems. But I noticed that it lowered the I/O efficiency of the containers. 
However, when I put in the .npmrc
`package-import-method=clone-or-copy`
All correct and with VirtioFS. 
Thanks!

**Comment by zkochan** [comment]
Closing: this is not something pnpm can fix.

Linking fails on a bind-mounted volume shared with the host. Keep node_modules inside the container, or use a named volume for it.

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
